lib/cretonne/src/loop_analysis.rs

```diff
@@ -129,77 +169,162 @@ impl LoopAnalysis {
-                            layout: &Layout) {
-        let mut stack: Vec<Ebb> = Vec::new();
+                            cfg: &mut ControlFlowGraph,
+                            domtree: &mut DominatorTree,
+                            func: &mut Function) {
+        let mut stack: Vec<(Ebb, Inst)> = Vec::new();
         // We handle each loop header in reverse order, corresponding to a pesudo postorder
         // traversal of the graph.
         for lp in self.loops().rev() {
             for &(pred, pred_inst) in cfg.get_predecessors(self.loops[lp].header) {
-                // We follow the back edges
-                if domtree.dominates(self.loops[lp].header, pred_inst, layout) {
-                    stack.push(pred);
+                // We follow the back edges only
+                if domtree.dominates(self.loops[lp].header, pred_inst, &func.layout) {
+                    stack.push((pred, pred_inst));
                 }
             }
-            while let Some(node) = stack.pop() {
+            // Then we proceed to discover loop blocks by doing a "reverse" DFS
+            while let Some((ebb, loop_edge_inst)) = stack.pop() {
                 let continue_dfs: Option<Ebb>;
-                match self.ebb_loop_map[node].expand() {
-                    None => {
-                        // The node hasn't been visited yet, we tag it as part of the loop
-                        self.ebb_loop_map[node] = PackedOption::from(lp);
-                        continue_dfs = Some(node);
+                match self.ebb_loop_map[ebb] {
+                    EbbLoopData::NoLoop() => {
+                        // The ebb hasn't been visited yet, we tag it as part of the loop
+                        self.ebb_loop_map[ebb] = EbbLoopData::Loop {
+                            loop_id: lp,
+                            last_inst: loop_edge_inst,
+                        };
+                        // We stop the DFS if the block is the header of the loop
+                        if ebb != self.loops[lp].header {
+                            continue_dfs = Some(ebb);
+                        } else {
+                            continue_dfs = None;
+                        }
                     }
-                    Some(node_loop) => {
-                        // We copy the node_loop into a mutable reference passed along the while
-                        let mut node_loop = node_loop;
-                        // The node is part of a loop, which can be lp or an inner loop
-                        let mut node_loop_parent_option = self.loops[node_loop].parent;
-                        while let Some(node_loop_parent) = node_loop_parent_option.expand() {
-                            if node_loop_parent == lp {
+                    EbbLoopData::Loop {
+                        loop_id,
+                        last_inst: _,
+                    } => {
+                        // So here the ebb we're visiting is already tagged as being part of a loop
```

Comment:
More comment nits: period after this sentence. And if you reflow this whole paragraph it would be tidier.